core/src/main/scala/kafka/server/AbstractBrokerToControllerChannelManager.scala

```diff
@@ -38,28 +39,36 @@ trait BrokerToControllerChannelManager {
   def sendRequest(request: AbstractRequest.Builder[_ <: AbstractRequest],
                   callback: RequestCompletionHandler): Unit
 
+  def clusterId(): Future[String]
+
   def start(): Unit
 
   def shutdown(): Unit
 }
 
 
 /**
- * This class manages the connection between a broker and the controller. It runs a single
- * {@link BrokerToControllerRequestThread} which uses the broker's metadata cache as its own metadata to find
- * and connect to the controller. The channel is async and runs the network connection in the background.
- * The maximum number of in-flight requests are set to one to ensure orderly response from the controller, therefore
- * care must be taken to not block on outstanding requests for too long.
+ * This abstract class manages the connection between a broker and the controller. It runs a single
+ * {@link BrokerToControllerRequestThread} which finds and connects to the controller. The channel is async and runs
+ * the network connection in the background. The maximum number of in-flight requests are set to one to ensure orderly
+ * response from the controller, therefore care must be taken to not block on outstanding requests for too long.
  */
-class BrokerToControllerChannelManagerImpl(metadataCache: kafka.server.MetadataCache,
-                                           time: Time,
-                                           metrics: Metrics,
-                                           config: KafkaConfig,
-                                           threadNamePrefix: Option[String] = None) extends BrokerToControllerChannelManager with Logging {
-  private val requestQueue = new LinkedBlockingDeque[BrokerToControllerQueueItem]
-  private val logContext = new LogContext(s"[broker-${config.brokerId}-to-controller] ")
-  private val manualMetadataUpdater = new ManualMetadataUpdater()
-  private val requestThread = newRequestThread
+abstract class AbstractBrokerToControllerChannelManager(time: Time,
```

Comment:
I think we need to rethink this part.  Deep inheritance hierarchies make the code difficult to understand.  Here we have the control flow bouncing between BrokerToControllerManager, AbstractBrokerToControllerManager, and Kip500BrokerToControllerManager, which makes the code hard to understand.

It seems like we should just look at the MetadataCache to see who the current controller is (if there is one).  Then we could use the same code for the pre and post-KIP-500 cases.

Another thing we need to do here is hook up `MetaLogManager` to the `MetadataCache`, so that when the leader changes, that change is reflected in the cache.  cc @hachikuji

Reply:
I filed KC-1265 to follow up on this.